### DISCUSSION TITLE
why windterm is so slow ?

### DISCUSSION THREAD

**Discussion by valmont2k** [kingToolbox/WindTerm#42] [Q&A]
i work on windows 10 and using WindTerm_Win10_Conpty.exe . Windterm is fine, but when i work in midnight commander in linux session it so slow. I noticed that moving up and down mc is slow and even the context menu is slow. Nice, promising terminal, but I can't work in it yet. I attach a video in comparison with another terminal on the same remote machine

https://user-images.githubusercontent.com/16284267/103643474-77e83380-4fa0-11eb-9f9d-dc545105799f.mp4

**Comment by kingToolbox** [comment]
Thank you and the video.

Although I have been working hard to improve performance and be satisfied with it, I have to say that what you mentioned is correct. After seeing your post, I have spent two or three hours testing several scenarios. One of them is keep pressing the down button in mc to traverse a folder with 168 items from top to bottom. WindTerm took 8.6 seconds, while MobaXterm/PuTTY only took 6.4 seconds (MobaXterm is based on PuTTY). It is worth mentioning that if I connect the power adapter to laptop, then WindTerm will take 7.3 seconds, while MobaXterm/PuTTY will still take 6.3 seconds.

The above facts illustrate two points. One is that the processing is indeed slow, and the other is that WindTerm requires more computing resources.

Compared with MobaXterm/Putty, WindTerm will try to do the extra following tasks when browsing:

- Update margins: line number and timestamp of each line, and highlight the fold range.
- Search and highlight the word where the caret is located in current session.
- Analyze the highlighting, folding and outline of the new text if the scheme is not None (Default scheme is Linux).

For the right mouse button menu, the mouse message is also sent to the server. So you can see that while the menu is displayed, the item under the mouse in mc will be selected, but MobaXterm/PuTTY does not do this, they only transmit mouse click and double click events to the server for processing.

But I don't think that these extra work will cause performance so much slower. I profiled and measured the performance of each function and found that most of the time was spent on painting. I will try to improve the performance in v2.0 to at least the same as MobaXterm/PuTTY.

In addition, double-clicking is not handled correctly in mc, so it will be fixed too.

**Reply by valmont2k** [reply]
thank you very much for the quick response. Once again I want to point out that I'm really impressed with windterm, impressed with its looks, the idea of ​​a fast application in the age of slow software. Built-in access to files of the target machine, excellent syntax highlighting, great fonts and completely free use - this is something that is not on the market now and has already sent out links to your site to my friends. But when I was going to switch to it, everything that you described - the slow transition inside mc and the slowness of rendering add frustration to the user experience. The impression is created that the software requires additional efforts from you, this is inconvenient for me. I am looking forward to the development of the project, as soon as the rendering is comparable to other software, I will switch to windterm. Thank you for such a necessary and interesting software, I wish the project prosperity.

**Reply by kingToolbox** [reply]
Thank you for recommending WindTerm to your friends and hope WindTerm can satisfy you and your friends :smile: 

I have good news for you. After experimenting, I found that if WindTerm do not draw the text or draw only a few characters at the beginning of the line, the rendering speed will be almost as fast as MobaXterm/PuTTY. So I think this is the main reason for the slowness and I think I know how to improve it easily.

At present, the text will be drawn character by character. If there are 80 characters in a line, the function **drawText** will be called 80 times. When the up key or down key are continuously pressed in mc, all lines need to be updated, and this drawing strategy greatly reduces the drawing performance. Obviously, we only need to output text with the same attributes at one time, and the speed will increase linearly. For example, now we need to output `hello world`, `hello` is bold blue and `world` is red, then the old drawing strategy needs to call **drawText** 11 times to draw each character, while the new one only needs to call **drawText** three times to draw `hello`, `space` and `world`. The speed can be almost tripled.

Although the new drawing strategy has not been implemented yet, I think it should be feasible. Of course, the final effect will not be known until I realize it in v2.0. I expect v2.0 will be released soon in mid-January.

It should be mentioned that the old drawing strategy is not a design mistake. The core of WindTerm is a high-performance text editor, I call it [WindEdit](https://github.com/kingtoolbox/digedit/). One feature of WindEdit is that it can support very long lines. For example, you can [view and edit a 400MB or even longer line without wrap](https://kingtoolbox.github.io/2020/01/21/400m-long-line-file/). In order to accurately scroll to the horizontal offset of each character, The line text must be accurately drawn character by character, instead of drawing the text in segments. Because when you output text in batches, glyph metrics may do not add up anymore when considering kerning, even for Latin text. In other words, the assumption that width('a') + width('b') is equal to width("ab") is not necessarily true. Although the difference between the two is only a few pixels when the text is short, but when the text is hundreds of megabytes, the difference is enough to make it impossible to locate any character correctly. In addition, font ligatures, etc. can also cause drawing deviations. So the drawing strategy must be designed this way.

For terminal, the old drawing strategy is not applicable. One is to reduce the speed, and the other is that usually only hundreds of characters need to be drawn in a line, and the line is also wrapped by default. So the new and faster strategies can be applied.

By the way, I artificially limit the refresh rate of the screen text. Under normal circumstances, 60 frames per second. Usually the system like Windows 10 supports 60 frames per second by default, so a higher refresh rate is more of a gimmick and can also reduce battery life because it uses more power. I think 60Hz is good enough.

**Reply by valmont2k** [reply]
good time of day
it's just great that the 2.0 release is coming so soon!
I also noticed that the rendering speed depends on the processor frequency. The fact is that through the power saving parameters I limit the processor frequency of my laptop to 0.8 GHz and in this state, slow rendering, the windterm is clearly visible. I do this because I noticed that this way the laptop heats up much less and the fans of my laptop are much less noisy, which interferes with work less. In addition, at the lower frequency, the battery life of the notebook is two to three times longer than at the normal frequency.
I don't know how technically it will be expensive, but it might be worth moving the interface rendering method into a separate setting and making a "fast" rendering, which will allow setting this setting to get high responsiveness of the program, even, perhaps, due to simplification elsewhere. such as using a monospaced font or disabling support for extra-long lines. It seems to me that working with huge text files is not as common in a typical work as the speed of work.
I repeat once again that you and only you decide where the project will move and what decisions to make in it, for my part, I just voice my wishes and in no case am I going to choose or demand for you.
Thanks again for your work.

**Reply by kingToolbox** [reply]
I am very happy to listen to your suggestions at any time, and your suggestions are a good supplement to me. Designing this as a setting is a good idea, especially for expert users. Although I hope that the default configuration can satisfy every user, this is usually unrealistic, so I think I will add the option if necessary.

Today I almost finished **infinite scrollback**. By dynamically compressing and recycling memory, WindTerm can support billions of lines of text with very little memory. I tested a 70 million lines, 10GB scrolling history, and the results looked pretty good. So I think I should have time to deal with the rendering speed now, if all goes well, I hope to be able to make a new prerelease for you this weekend, and also hope the improved performance can meet your expectations.

**Comment by kingToolbox** [comment]
Sorry to keep you waiting. It took time to improve some trivial functions.

I just released [WindTerm_2.0.0_Prerelease_1](https://github.com/kingToolbox/WindTerm/releases/tag/2.0.0), and I believe that the rendering speed is still a bit slower due to the complexity of painting, but it is almost comparable to MobaXterm/PuTTY, please download and check. Thank you.

**Comment by valmont2k** [comment]
firstly, moving through the lines of the midnight commander has become really fast. It's nice. As for the appearance of the context menu, I could not see that something has changed from the previous time. I would like to praise you for the "copy and paste" context menu item, it is sorely lacking in other ssh clients. Taking this opportunity, I want to know if it is even possible to copy and paste text in the console via "ctrl + c ctrl + v"?

Just tried it, really ctrl + shift + x works
this is really just what I need and faster than using the context menu. In this case, I will roll back to the previous version (where tab works) and work without the context menu.

Yes, I forgot to mention, in the prerelease, the tab did not work for me (on one of the 2 servers where I tried it), apparently, it broke, everything is reflected in the video.

**Reply by kingToolbox** [reply]
I am glad you like `copy and paste`. And I also recommend that you use shortcut keys to operate, which is more convenient and faster than using the menu. In addition to "ctrl+shift+x", you can also use "ctrl+shift+c" to copy and "ctrl+shift+v" to paste. You can find all shortcuts in `WindTerm_2.0.0\global\wind.keymaps` or see [frequently used shortcut keys](https://github.com/kingToolbox/WindTerm#shortcuts).

There is also a little trick in WindTerm, you can quickly input by [dragging and dropping the selected text](https://kingtoolbox.github.io/2020/06/20/tips-input-by-dragging-and-dropping-text/) and more tips are [here](https://kingtoolbox.github.io/tags/tips/).

As for ctrl+c and ctrl+v, because the server needs to use them, so they cannot be used for copying or pasting. If you really want to use it, you can press `shift+enter` to enter the local mode, and then you can use them at will. For more detailed local mode information, please read [local mode](https://kingtoolbox.github.io/tags/local-mode/).

The tab error really stumped me. The related code and configuration have not been changed this time, and I also tested it on multiple systems just now and the tab function is normal.

If it is convenient for you, before entering the tab, you can click the menu item `Mode - Debug Mode` in the v1.9.2 and v2.0.0_prerelease_1 respectively, and click the menu item again after entering multiple tabs. Then you will get two logs in`WindTerm_1.9.2\profiles\default.v10\terminal\logs\debug` and `WindTerm_2.0.0\profiles\default.v10\terminal\logs\debug`. You can use a comparison tool to see the difference, or send the log to me for analysis. Or you can also try whether the tab key is normal in cmd or powershell. So far, I have not encountered the tab key not working.

v2.0.0_Prelease_1 only improve the rendering speed of lines, the speed of the menu still needs to be analyzed. By the way, do you think the right-click menu is slow when you are not in mc?

BTW, I deleted your video because it contains information such as your server address. I hope you don't mind.

**Reply by kingToolbox** [reply]
I analyzed this again and observed that the tab key was indeed sent and parsed by the server, but it did not trigger the auto-completion and treated it as a real '\t'. So I speculate whether this is not WindTerm's problem, but Shell's problem.

But you said that it is good when using v1.9.2. So what about other clients, such as Mobaxterm/PuTTY? It's really strange.
